llpc/patch/llpcPatchBufferOp.cpp

```diff
@@ -1564,512 +1585,279 @@ Value* PatchBufferOp::ReplaceLoad(
+
+    Value* pStoreValue = nullptr;
+    if (!isLoad)
+    {
+        pStoreValue = pStoreInst->getValueOperand();
+
+        if (pStoreValue->getType()->isPointerTy())
+        {
+            pStoreValue = m_pBuilder->CreatePtrToInt(pStoreValue, m_pBuilder->getIntNTy(bytesToHandle * 8));
+            CopyMetadata(pStoreValue, pStoreInst);
+        }
+
+        pStoreValue = m_pBuilder->CreateBitCast(pStoreValue, pCastType);
+        CopyMetadata(pStoreValue, pStoreInst);
+    }
 
-    uint32_t remainingBytes = bytesToLoad;
+    // The index in pStoreValue which we use next
+    uint32_t storeIndex = 0;
 
+    uint32_t remainingBytes = bytesToHandle;
     while (remainingBytes > 0)
     {
-        const uint32_t offset = bytesToLoad - remainingBytes;
+        const uint32_t offset = bytesToHandle - remainingBytes;
         Value* pOffset = (offset == 0) ?
                                pBaseIndex :
                                m_pBuilder->CreateAdd(pBaseIndex, m_pBuilder->getInt32(offset));
 
-        Type* pIntLoadType = nullptr;
-        uint32_t loadSize = 0;
+        Type* pIntType = nullptr;
+        uint32_t size = 0;
 
-        // Load the greatest possible size
+        // Handle the greatest possible size
         if (alignment >= 4 && remainingBytes >= 4)
         {
             if (remainingBytes >= 16)
             {
-                pIntLoadType = VectorType::get(Type::getInt32Ty(*m_pContext), 4);
-                loadSize = 16;
+                pIntType = VectorType::get(Type::getInt32Ty(*m_pContext), 4);
+                size = 16;
             }
             else if (remainingBytes >= 12 && !isInvariant)
             {
-                pIntLoadType = VectorType::get(Type::getInt32Ty(*m_pContext), 3);
-                loadSize = 12;
+                pIntType = VectorType::get(Type::getInt32Ty(*m_pContext), 3);
+                size = 12;
             }
             else if (remainingBytes >= 8)
             {
-                pIntLoadType = VectorType::get(Type::getInt32Ty(*m_pContext), 2);
-                loadSize = 8;
+                pIntType = VectorType::get(Type::getInt32Ty(*m_pContext), 2);
+                size = 8;
             }
             else
             {
                 // remainingBytes >= 4
-                pIntLoadType = Type::getInt32Ty(*m_pContext);
-                loadSize = 4;
+                pIntType = Type::getInt32Ty(*m_pContext);
+                size = 4;
             }
         }
         else if (alignment >= 2 && remainingBytes >= 2)
         {
-            pIntLoadType = Type::getInt16Ty(*m_pContext);
-            loadSize = 2;
+            pIntType = Type::getInt16Ty(*m_pContext);
+            size = 2;
         }
         else
         {
-            pIntLoadType = Type::getInt8Ty(*m_pContext);
-            loadSize = 1;
+            pIntType = Type::getInt8Ty(*m_pContext);
+            size = 1;
         }
-        LLPC_ASSERT(loadSize != 0);
+        LLPC_ASSERT(size != 0);
```

Comment:
This should also check that `pIntType` is not null, or the base case when size=1 and type=i8 could be moved to the top.